mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java

```diff
@@ -465,20 +467,27 @@ public void setEnableSuffixMatch(boolean suffixMatch) {
     }
 
     @Override
-    public List<MessageId> search(MailboxSession session, Collection<MailboxId> mailboxIds, SearchQuery searchQuery, long limit) throws MailboxException {
+    public MessageResults search(MailboxSession session, Collection<MailboxId> mailboxIds, SearchQuery searchQuery, long limit) throws MailboxException {
+        List<MessageId> allMessageInMailboxes = searchAllMessageInMailboxes(session, mailboxIds, searchQuery).collect(Guavate.toImmutableList());
+        return new MessageResults(allMessageInMailboxes.size(),
+            allMessageInMailboxes
+                .stream()
+                .limit(Long.valueOf(limit).intValue())
+                .collect(Guavate.toImmutableList()));
+    }
+
+    private Stream<MessageId> searchAllMessageInMailboxes(MailboxSession session, Collection<MailboxId> mailboxIds, SearchQuery searchQuery) throws MailboxException {

```

Comment:
Mmmmmmh
That don't look efficient. This is for production use! We should find a way to do it well. use TopDocs.totalHits + cardinality aggregation like it's done in ES.... (ES is built on top of Lucene so doing this is maybe not easy but clearly doable)

Reply:
I tried but I cannot find the same way with lucene :(. Please see my latest commit

Reply:
Update lucene to 4.10.4.

Rely on `DuplicateFilter` in lucene-sandbox project.

I don't really care about migration as we can say, in case of data storage changes `just reindex.`